### PR TITLE
Peterson_Fix_Weekly_Summaries_edit_Last_week

### PR DESCRIPTION
**Pull request by peterson337**
# Description
In the HGN software, specifically within the 'Tasks and Timelogs' card on the Dashboard, there is a tab named 'Weekly Summaries,' which displays three editable summaries. Upon clicking the 'edit' button below each summary, two additional buttons, 'save' and 'cancel,' appear. The bug was found in the 'save' button functionality, which, when triggered, initiated a 'put' request resulting in two errors. The first error occurred because Node did not receive information from React, resulting in an HTTP status code 500, indicating a server error. The second error was related to the frontend-to-backend request: the request URL should have contained the user ID but was instead sent as 'undefined.'

## Related PRS (if any):
This frontend PR is associated with the following backend PR: https://github.com/OneCommunityGlobal/HGNRest/pull/781

## Main changes explained:
The WeeklySummaries component has been modified to fix the bugs.

## How to test:
1. check into current branch
2. do `npm install` and `npm start` to run this PR locally
3. log as any user.
4. Go to Dashboard →  Weekly Summaries.
5. Click the "Edit" button.
6. After editing the desired summary, click the "Save" button.
7. Open the browser's Developer Tools by pressing F12 or right-clicking and selecting 'Inspect.' Navigate to the 'Console' tab. Check if the following errors appear in the console: a  
8. If the errors shown in the image above do not appear and the summary has been edited successfully, reload the page and check if the summary remains edited.

## Screenshots or videos of changes:

Peterson_Fix_Weekly_Summaries_edit_Last_week’s

## Note:
The console is going to have some errors not related to this PR. These can be disregarded because the code changed by this PR will not introduce these errors to the "development" branch when it gets merged.

### CODE REVIEW

**Review by meetpadhiar4 — DISMISSED** [review]
Hi, I tested your PR and it works as intended. I was able to edit the summary without any errors and the communication between the server and the frontend is happening properly resulting in 200 status code. Refer the video attached for reference.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/67577022/2ae5da1f-257d-481c-9df2-0fe1b9ae68ec

**Comment by nnamdixobi** [comment]
testt

Everything compiles as expected but for some reason it wont let me save a weekly summary for testing

**Review by psharma1984 — DISMISSED** [review]
Hi @peterson337 I reviewed and its working as expected. Able to edit the summary and its saved and remains updated when page is refreshed
a
b

**Review by hetvi11110 — DISMISSED** [review]
I tested this PR as mentioned. It is working fine on my local server. Here is the screenshot.
PR2023

**Review by Tiantian-C — DISMISSED** [review]
I have tested the function,and it works well!

File: src/actions/weeklySummaries.js [hunk 1]
```diff
@@ -79,7 +79,8 @@ export const updateWeeklySummaries = (userId, weeklySummariesData) => {
 
       // Merge the weekly summaries related changes with the user's profile.
       const {mediaUrl, weeklySummaries, weeklySummariesCount } = weeklySummariesData;
-      console.log('respon get', response.data)
+      
+      // console.log('respon get', response.data)
```
Comment (Abi-Liu) [inline src/actions/weeklySummaries.js:83]: Remove console.log

**Review by Abi-Liu — CHANGES_REQUESTED** [review]
Hey! Overall the code works as expected. However, there is one small bug I encountered. When editing the weekly summaries, if a user tries to save an invalid summary (either blank or less than 50 words), the error will fire as expected but the button will stay disabled. This forces the user to refresh and then make the changes again before being allowed to update.

This can be fixed by just setting the `LoadingHandleSave` state to null if an error occurs.

EDIT: Also no need to await `setLoadingHandleSave` on line 92 in `src/components/Timelog/WeeklySummaries.jsx`

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/59585815/875035fb-e91a-433e-af57-b946ed2edb1e

**Comment by peterson337** [comment]
> Hey! Overall the code works as expected. However, there is one small bug I encountered. When editing the weekly summaries, if a user tries to save an invalid summary (either blank or less than 50 words), the error will fire as expected but the button will stay disabled. This forces the user to refresh and then make the changes again before being allowed to update.
> 
> This can be fixed by just setting the `LoadingHandleSave` state to null if an error occurs.
> 
> EDIT: Also no need to await `setLoadingHandleSave` on line 92 in `src/components/Timelog/WeeklySummaries.jsx`
> 
>  video1131734987.mp4



> Hey! Overall the code works as expected. However, there is one small bug I encountered. When editing the weekly summaries, if a user tries to save an invalid summary (either blank or less than 50 words), the error will fire as expected but the button will stay disabled. This forces the user to refresh and then make the changes again before being allowed to update.
> 
> This can be fixed by just setting the `LoadingHandleSave` state to null if an error occurs.
> 
> EDIT: Also no need to await `setLoadingHandleSave` on line 92 in `src/components/Timelog/WeeklySummaries.jsx`
> 
>  video1131734987.mp4

I fixed the bug. See my gif for more reference:
Video 1

File: src/actions/weeklySummaries.js (hunk 1 above)
Comment (peterson337) [reply src/actions/weeklySummaries.js]: I removed the console.log from the weeklySummaries.js archive.

photo 4

**Review by Abi-Liu — CHANGES_REQUESTED** [review]
Just an extra set state that is setting the state to the same value. Other than that the changes look good!

File: src/components/Timelog/WeeklySummaries.jsx [hunk 2]
```diff
@@ -48,31 +64,39 @@ const WeeklySummaries = ({ userProfile }) => {
...
+    
+    // This code saves edited weekly summaries in MongoDB.
+    await dispatch(updateWeeklySummaries(userProfile._id, updatedUserProfile));
     await dispatch(getUserProfile(userProfile._id));
-      // Toggle off editing mode
-      toggleEdit(index);
-    } else {
-      // Invalid summary, show an error message or handle it as needed
-      alert('Please enter a valid summary with at least 50 words.');
+    await setLoadingHandleSave(null);
```
Comment (Abi-Liu) [inline src/components/Timelog/WeeklySummaries.jsx:92]: Delete this line. Sorry I quoted the wrong code snippet in my other comment

Comment (peterson337) [reply src/components/Timelog/WeeklySummaries.jsx:92]: Hi [Abi-Liu](https://github.com/Abi-Liu).  I removed the setLoadingHandleSave from the handleSave function.

photo 3

**Comment by SanketKaware** [comment]
Followed all the mentioned steps and the changes look good to me. Those errors didn't appear.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/17793708/297a191e-2651-4aad-912c-fea7dc2deec0

**Review by Abi-Liu — DISMISSED** [review]
Looks good!

File: src/components/Timelog/WeeklySummaries.jsx (hunk 2 above)
Comment (Abi-Liu) [reply src/components/Timelog/WeeklySummaries.jsx:92]: It's still there. Underneath `await dispatch(getUserProfile(userProfile._id));`

**Review by 20chen-7 — COMMENTED** [review]
npm run locally, the function looks good. but when I run npm start, npm test errors occur.
Screenshot 2024-03-15 at 4 34 31 PM
Screenshot 2024-03-15 at 4 34 42 PM
Screenshot 2024-03-15 at 4 22 40 PM

**Review by SanketKaware — DISMISSED** [review]

**Review by SRamen1999 — COMMENTED** [review]
This PR works as intended. I was able to edit the summary without any errors and the communication between the server and the frontend is happening properly resulting in 200 status code.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/56932664/1160bc59-3ff6-4ba3-a336-06da0a30a8da

There is also a safe check to see if the user edits the summary less than 50 words. 

Screen Shot 2024-03-15 at 5 44 26 PM

**Comment by rajnada** [comment]
Hii,

I have tested the functionality and it is working fine!
Good Job

Before :




After :

**Review by PratimaSingh02 — DISMISSED** [review]
Hi I took the following steps to execute this PR:
1. Logged in as an owner
2. Created a test weekly summary
3. Edited the weekly summary
4. Verified that the server error was not there
5. Reloaded the page to check if the edit was still present
image

**Review by sameersdeshpande — DISMISSED** [review]
Hello @peterson337 
The changes are working as expected. I checked in to current branch and logged in as a user and edited the summary and found no errors in console. It was with 200 ok status.
Thanks
Screenshot 2024-03-16 at 7 12 31 PM
Screenshot 2024-03-16 at 7 12 48 PM
Screenshot 2024-03-16 at 7 13 00 PM

**Review by iven-yao — DISMISSED** [review]
tested the edit function in weekly summaries, and it works perfect without any error in console, and reflect the editing after save correctly.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/25358966/96b4d174-6b93-413d-8c78-80ef422fa22c

File: src/components/Timelog/WeeklySummaries.jsx [hunk 3]
```diff
@@ -1,23 +1,40 @@
...
   const [editing, setEditing] = useState([false, false, false]);
+
   const [editedSummaries, setEditedSummaries] = useState([
     userProfile.weeklySummaries[0]?.summary || '',
     userProfile.weeklySummaries[1]?.summary || '',
     userProfile.weeklySummaries[2]?.summary || '',
   ]);
   const [originalSummaries, setOriginalSummaries] = useState([...editedSummaries]);
 
+  const [LoadingHandleSave, setLoadingHandleSave] = useState(null);
```
Comment (ImzIssa) [inline src/components/Timelog/WeeklySummaries.jsx:34]: // Consider refactoring to use camelCase for consistency everywhere in the codebase.
const [loadingHandleSave, setLoadingHandleSave] = useState(null);

**Review by ImzIssa — DISMISSED** [review]
Hey, I've tested this PR and the functionality seems to be implemented correctly. Nice work!
Screenshot (32)

I do have a minor suggestion to enhance the UI. Consider using the reactstrap Alert inplace of the browser alert.
Screenshot (33)

the alert when creating a summary

Screenshot (34)

**Review by mohabbasd — DISMISSED** [review]
Hello, I tested your PR and it works perfectly. Below is a video for reference.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91215370/555b0919-e661-4c9e-b75b-1108e4f60fa1

**Review by Coops023 — DISMISSED** [review]
Hi i reviewed this PR and was able to edit summaries with out an error! good job.

test

**Comment by anandsesha** [comment]
Hi, I have tested this functionality with its associated backend PR #781, logging in as admin user and volunteer user.

- Created weekly summary and while editing that summary checked console and the mentioned error is not seen.
- After saving the edit, inspected Network tab which results in status 200 Ok as expected.
- Reloaded page and Edited summary is persisted.
- Issue reported earlier of summary less than 50 is not allowed. Appropriate message shown.

Below screenshots for reference:

PR 2023 - Edited Network tab
PR 2023 - Info persisted
PR 2023 - No console error
PR 2023 - Pre-edit
PR 2023 - summary less 50

**Review by Sarthak32 — COMMENTED** [review]
Hi , I tested your PR. Everything works as expected but for some reason it does not let me save my edited summary.
Screenshot 2024-03-24 121544
Screenshot 2024-03-24 121555

**Review by Jaiwin-4998 — DISMISSED** [review]
Hi, I have tested you PR and it works properly. The description remains the edited after the changes and no errors.
PR 2023

**Review by kylepham — DISMISSED** [review]
Everything works as expected. New summary edit was successfully saved and returned 200 status code. 50-word guard also worked.

Before: 

Screenshot 2024-03-27 at 2 52 44 PM (2)

After: 

Screenshot 2024-03-27 at 2 53 30 PM (2)


Screenshot 2024-03-27 at 2 54 27 PM (2)

**Comment by rajnada** [comment]
Hii,

I have tested the PR. It works completely fine! Good Job.

image

**Comment by MQChen211** [comment]
Hi, I tested your PR and it works as expected.
PR#2023+781

**Review by aaryaneil — APPROVED** [review]
Tested the PR and there was no error generated while editing summary.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/39597927/61df8d27-730b-43df-98c6-e7713b7a27d3
Screenshot 2024-04-20 at 6 15 15 PM

**Comment by one-community** [comment]
Thank you all, merging!

**Comment by one-community** [comment]
Thank you all, merging!

**Pull request merged**